core/src/main/java/hudson/model/Job.java

```diff
@@ -675,6 +677,40 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
         Util.deleteRecursive(getBuildDir());
     }
 
+    @Restricted(NoExternalUse.class)
+    @Extension
+    public static class SubItemBuildsLocationImpl extends ItemListener {
+        @Override
+        public void onLocationChanged(Item item, String oldFullName, String newFullName) {
+            final Jenkins jenkins = Jenkins.getInstance();
+            if (!jenkins.isDefaultBuildDir() && item instanceof Job) {
```

Comment:
Yes, it already behaves correctly for the default directory, i.e. when the builds are somewhere under ITEM_ROOTDIR, so this is to make sure we only do this extra check when we have something different.